atlasdb-commons/src/main/java/com/palantir/common/concurrent/ConcurrentStreams.java

```diff
@@ -56,36 +54,38 @@ private ConcurrentStreams() {}
             return values.stream().map(mapper);
         }
 
-        Map<T, CompletableFuture<S>> futuresByValue = values.stream()
-                .collect(Collectors.toMap(Function.identity(), value -> new CompletableFuture<>()));
-        Queue<T> valueQueue = new ConcurrentLinkedQueue<T>(values);
+        List<StreamElement<T, S>> elements = values.stream()
+                .map((Function<T, StreamElement<T, S>>) StreamElement::new)
+                .collect(Collectors.toList());
+        Queue<StreamElement<T, S>> queue = new ConcurrentLinkedQueue<>(elements);
 
         if (size < concurrency) {
             concurrency = size;
```

Comment:
assigning value to a parameter feels odd.

Reply:
Okay how about something like:
```
int internalConcurrency = size < concurrency ? size : concurrency;
```